streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1271,4 +1279,81 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
         builder.addGraphNode(streamsGraphNode, processNode);
     }
+
+    @Override
+    public KTable<K, V> toTable() {
+        return toTable(NamedInternal.empty());
+    }
+
+    @Override
+    public KTable<K, V> toTable(Named named) {
+        return doToTable(named, null);
+    }
+
+    @Override
+    public <VR> KTable<K, VR> toTable(Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        return toTable(NamedInternal.empty(), materialized);
+    }
+
+    @Override
+    public <VR> KTable<K, VR> toTable(Named named,
+                                      Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized) {
+        Objects.requireNonNull(materialized, "materialized can't be null");
+
+        final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal =
+            new MaterializedInternal<>(materialized);
+        return doToTable(named, materializedInternal);
+    }
+
+    private <VR> KTable<K, VR> doToTable(Named named,
+                                         MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal) {
+        Objects.requireNonNull(named, "named can't be null");
+        final String name = new NamedInternal(named).orElseGenerateWithPrefix(builder, TO_KTABLE_NAME);
+        final Serde<K> keySerde;
+        final Serde<VR> valueSerde;
+        final String queryableStoreName;
+        final String storeName;
+        final StoreBuilder<TimestampedKeyValueStore<K, VR>> storeBuilder;
+
+        if (materializedInternal != null) {
+            keySerde = materializedInternal.keySerde() != null ? materializedInternal.keySerde() : this.keySerde;
+            valueSerde = materializedInternal.valueSerde();
+            queryableStoreName = materializedInternal.queryableStoreName();
+            storeBuilder = materializedInternal.queryableStoreName() == null
+                ? null
+                : new TimestampedKeyValueStoreMaterializer<>(materializedInternal).materialize();
+
+            storeName = materializedInternal.storeName() == null
+                ? builder.newStoreName(TO_KTABLE_NAME)
+                : materializedInternal.storeName();
+
+        } else {
+            keySerde = this.keySerde;
+            storeName = builder.newStoreName(TO_KTABLE_NAME);
+            valueSerde = null;
+            queryableStoreName = null;
+            storeBuilder = null;
+        }
+
+        final KTableSource<K, VR> processorSupplier = new KTableSource<>(storeName, queryableStoreName);
+        final ProcessorParameters<K, VR> processorParameters = new ProcessorParameters<>(processorSupplier, name);
+
+        final StreamsGraphNode tableNode = new TableProcessorNode<>(
```

Comment:
Using `TableProcessorNode` implies that we make all decision about materialization in `doToTable`? -- However, I am wondering if this is correct. When we call `StreamsBuilder.table()`, we create a `TableSourceNode` (within `InternalStreamsBuilder#table()` that allows us to defer decision when we build the actual `Topology`. This has the advantage that we can take into account how the `KTable` is used downstream. For example, if we use it downstream in a stream-table join, we need to materialize the table -- but if we decide to non materialized here, we can't change this decision later on using `TableProcessorNode`? (Correct me if I am wrong.)

Reply:
Oh yeah! Thank you! I almost missed it.

Reply:
@mjsax 
TableProcessorNode can also be changed later.
Do not create stores when StoreNames and StoreBuilder are null !
 
TableProcessorNode.class
 public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
        final String processorName = processorParameters.processorName();
        topologyBuilder.addProcessor(processorName, processorParameters.processorSupplier(), parentNodeNames());

        if (storeNames.length > 0) {
            topologyBuilder.connectProcessorAndStateStores(processorName, storeNames);
        }

        // TODO: we are enforcing this as a keyvalue store, but it should go beyond any type of stores
        if (storeBuilder != null) {
            topologyBuilder.addStateStore(storeBuilder, processorName);
        }
    }

Reply:
@mjsax 
Thank you!
TableProcessorNode and TableSourceNode are not very different from TopologyBuilder.addSource() 

We may decide not to be realized before, but this decision can be changed later.
Even with TableProcessorNode It is possible.